File: Data Structures & Algorithms/path-with-minimum-effort/submission-6.cpp
class Solution {
public:
    int minimumEffortPath(vector<vector<int>>& heights) {
        // Time Complexity: O((rows*cols)log(rows*cols) Space Complexity: O(rows*cols))
        set<pair<int, int>> visited;
        set<pair<int, int>> moves{{0,1},{1,0},{0,-1},{-1,0}};
        priority_queue<tuple<int, int, int>, vector<tuple<int, int, int>>, greater<tuple<int, int, int>>> minHeap;
        minHeap.push({0,0,0});
        int rows{int(heights.size()) - 1}, cols{int(heights[0].size()) - 1};
        while (!minHeap.empty()) {
            auto [e, r, c] = minHeap.top();
            minHeap.pop();
            if (visited.contains({r, c})) continue;
            visited.insert({r, c});
            if (r == rows && c == cols) return e;
            for (auto [dr, dc] : moves) {
                int nr{r + dr}, nc{c + dc};
                if (min(nr, nc) >= 0 && nr < heights.size() && nc < heights[nr].size()) {
                    minHeap.push({max(e, abs(heights[r][c] - heights[nr][nc])), nr, nc});
                }
            }
        }
    }
};

File: Data Structures & Algorithms/shortest-path-in-binary-matrix/submission-6.py
class Solution:
    def shortestPathBinaryMatrix(self, grid: List[List[int]]) -> int:
        # Time Complexity: O(n^2) Space Complexity: O(n^2)
        moves = ((0,1),(1,0),(0,-1),(-1,0),(1,1),(1,-1),(-1,1),(-1,-1))
        visited = set()
        q = deque()
        length = 1
        rows = len(grid) - 1
        cols = len(grid[0]) - 1
        q.append([0,0])
        while q:
            layer = len(q)
            for i in range(layer):
                r, c = q.popleft()
                if grid[r][c] == 1: continue
                elif (r,c) == (rows, cols): return length
                for dr, dc in moves:
                    nr, nc = r + dr, c + dc
                    if min(nr, nc) >= 0 and nr < len(grid) and nc < len(grid[nr]) and (nr, nc) not in visited:
                        visited.add((nr, nc))
                        if grid[nr][nc] == 0: q.append([nr, nc])
            length += 1
        return -1